Add sane fallback for fee estimation

morcos opened the pull request:
Make fee estimation logic respect GetRequiredFee at all times.
In addition add new commandline option "-fallbackfee" to use when fee estimation does not have sufficient data.

MarcoFalke commented:
utACK  cc99c41

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -47,6 +47,12 @@ bool fSendFreeTransactions = DEFAULT_SEND_FREE_TRANSACTIONS;
  */
 CFeeRate CWallet::minTxFee = CFeeRate(DEFAULT_TRANSACTION_MINFEE);
+/**
+ * If fee estimation does not have enough data to provide estimates, the use this fee instead.
```
Nit: `, then use` ...

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -2223,14 +2229,12 @@ CAmount CWallet::GetMinimumFee(unsigned int nTxBytes, unsigned int nConfirmTarge
+            nFeeNeeded = fallbackFee.GetFee(nTxBytes);
+    }
+    // prevent user from paying a fee below minRelayTxFee or minTxFee
+    nFeeNeeded = std::max(nFeeNeeded, GetRequiredFee(nTxBytes));
```
This is important on it's own. Could you create another atomic commit with just this diff?

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -41,6 +41,8 @@ static const unsigned int DEFAULT_KEYPOOL_SIZE = 100;
 //! -paytxfee will warn if called with a higher fee than this amount (in satoshis) per KB
 static const CAmount nHighTransactionFeeWarning = 0.01 * COIN;
+//! -fallbackfee default
+static const CAmount DEFAULT_FALLBACK_FEE = 20000;
```
I'd say `11000` is enough, we can determine a "better" `DEFAULT_` for 0.13

morcos committed:
Always respect GetRequiredFee for wallet txs

morcos committed:
Add sane fallback for fee estimation

Add new commandline option "-fallbackfee" to use when fee estimation does not have sufficient data.

morcos commented:
Addressed @MarcoFalke's comments.
Although left fee = 20k satoshis until the bikeshedding is finished.  I don't care want goes there, but I propose using the result of estimatefee(4) with a significantly longer time decay over the last few months.  I'll post the result of that once calculated.

@wumpus this is meant for 0.12 backport

morcos commented:
@MarcoFalke ugh.. thanks.  more than just that one fail.  These are probably things that should be fixed b/c the travis tests that fail are implicitly depending on estimatefee returning minrelaytxfee, which isn't a good assumption.  I'm working on it

morcos committed:
SQUASHME: Fix rpc tests that assumed fallback to minRelayTxFee

wumpus commented:
Reminder: @wumpus on github is not a bitcoin person.

morcos commented:
Apologies.

@laanwj I think we could use this in 0.12

laanwj commented:
Concept ACK.

One nit: this adds yet another `-XXXfee` parameter.
Although it makes sense, for the wallet we already have `-mintxfee`, `-paytxfee`, `-maxtxfee` this wild growth of options does get a bit confusing for users, I think.
But I don't know a better solution either.

dcousens commented:
@laanwj its also not clear whether those parameters are referring to the wallet or to relay policy.

btcdrak commented:
Concept ACK will test later today.

sdaftuar commented:
Tested, ACK bebe58b748532157958f9055e4517e280684006c

MarcoFalke commented:
* utACK 995b9f3
* utACK bebe58b

laanwj commented:
utACK

laanwj commented:
Cherry-picked to 0.12 as a36d79bfe247e7fc5c6296fd8603f5094edfe558
